experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,957 @@
+      uint64_t res = lhsVal * rhsVal;
+      result.data->updateValue(num2vec(res, bitwidth));
+    }
+  }
+  void valuesActive() override {
+    llvm::outs() << getName() << ":\n";
+    lhs.valid->printValue();
+    llvm::outs() << "\n";
+    lhs.ready->printValue();
+    llvm::outs() << "\n";
+    lhs.data->printValue();
+    llvm::outs() << "\n";
+    rhs.valid->printValue();
+    llvm::outs() << "\n";
+    rhs.ready->printValue();
+    llvm::outs() << "\n";
+    rhs.data->printValue();
+    llvm::outs() << "\n";
+    result.valid->printValue();
+    llvm::outs() << "\n";
+    result.ready->printValue();
+    llvm::outs() << "\n";
+    result.data->printValue();
+    llvm::outs() << "\n";
+  }
+  std::string getName() override { return "MulI"; }
+  ~Muli() {}
+};
+class OEHB : public Component {};
+class Sink : public Component {
+public:
+  PortChannel ins;
+  void exec(bool clk) override { ins.ready->updateValue(true); }
+  void valuesActive() override {
+    llvm::outs() << getName() << ":\n";
+    ins.valid->printValue();
+    llvm::outs() << "\n";
+    ins.ready->printValue();
+    llvm::outs() << "\n";
+  }
+  std::string getName() override { return "Sink"; }
+
+  Sink() = default;
+  virtual ~Sink() {}
+};
+
+class Source : public Component {
+
+public:
+  PortChannel outs;
+  void exec(bool clk) override { outs.valid->updateValue(true); }
+  void valuesActive() override {
+    llvm::outs() << getName() << ":\n";
+    outs.valid->printValue();
+    llvm::outs() << "\n";
+    outs.ready->printValue();
+    llvm::outs() << "\n";
+  }
+  std::string getName() override { return "Source"; }
+
+  Source() = default;
+  virtual ~Source() {}
+};
+
+class Return : public Component {
+  unsigned dataWidth;
+
+public:
+  PortChannel ins, outs;
+  Return(unsigned tempDatawidth) : dataWidth(tempDatawidth) {}
+  void exec(bool clk) override {
+    ins.ready->updateValue(true);
+    outs.valid = ins.valid;
+    outs = ins;
+  }
+
+  void valuesActive() override {
+    llvm::outs() << getName() << ":\n";
+    ins.valid->printValue();
+    llvm::outs() << "\n";
+    ins.ready->printValue();
+    llvm::outs() << "\n";
+    ins.data->printValue();
+    llvm::outs() << "\n";
+    outs.valid->printValue();
+    llvm::outs() << "\n";
+    outs.ready->printValue();
+    llvm::outs() << "\n";
+    outs.data->printValue();
+    llvm::outs() << "\n";
+  }
+  std::string getName() override { return "Extsi"; }
+  ~Return() {}
+};
+
+/// State of the particular channel
+/// x.first == 1 => we have wire x (e.g valid)
+/// x.second == 0 => x is signal; x.second == 1 => x is data
+class State {
+public:
+  std::pair<bool, bool> valid;
+  std::pair<bool, bool> ready;
+  std::pair<bool, bool> data;
+};
+
+/// Represents circuit
+class Circuit {
+  // fucOp, that initializes the circuit
+  handshake::FuncOp funcOp;
+
+public:
+  /// This map contains the set of channels (so, it's basically the circuit
+  /// state)
+  llvm::DenseMap<mlir::Value, Channel> circuitState;
```

Comment:
While it is sort of ok-ish for now, ultimately this map should be from `mlir::Value` to `llvm::Any`, because there may be SSA operands in the circuit which are not dataflow channels.